### DISCUSSION TITLE
Getting I/O Car Model and Revision

### DISCUSSION THREAD

**Discussion by rpabech** [ottowayi/pycomm3#247] [Q&A]
Hi, by running the Discovery and LogixDriver I can get most data from my PLCs, but still having issues getting data from the IO cards on the slots.

When I use the:

	driver = CIPDriver(IPADDRESS)
	driver.open()	
	SlotInfo = driver.get_module_info(slot)

I get empty information (even if a slot has an IO card. See below:

{'vendor': 'Rockwell Automation/Allen-Bradley', 'product_type': 'General Purpose Discrete I/O', 'product_code': 82, 'revision': {'major': 3, 'minor': 1}, 'status': b'a\x00', 'serial': '00000000', 'product_name': ''}
{'vendor': 'Rockwell Automation/Allen-Bradley', 'product_type': 'General Purpose Discrete I/O', 'product_code': 85, 'revision': {'major': 3, 'minor': 1}, 'status': b'a\x00', 'serial': '00000000', 'product_name': ''}

I get different product codes, but the Product Name, Serial and Product Type do not seem correct.

Also using LogixDrive, I also do not get the info (it get the modules and slots, but do not get the IO Card model and version). It only shows the types. See below. Any idea on how to get this info on this structure? I really need to find out what Cards I have and where they are in the structure. Thanks in advance!!!

    "modules": {
        "New_INX_IO": {
            "slots": {
                "3": {
                    "types": [
                        "C"
                    ]
                },
                "2": {
                    "types": [
                        "O"
                    ]
                },
                "5": {
                    "types": [
                        "I"
                    ]
                },
                "6": {
                    "types": [
                        "C"
                    ]
                },
                "4": {
                    "types": [
                        "I"
                    ]
                },
                "1": {
                    "types": [
                        "I"
                    ]
                }
            }
        },
        "Local": {
            "slots": {
                "11": {
                    "types": [
                        "C"
                    ]
                },
                "4": {
                    "types": [
                        "I"
                    ]
                },
                "13": {
                    "types": [
                        "O"
                    ]
                },
                "12": {
                    "types": [
                        "O"
                    ]
                },
                "6": {
                    "types": [
                        "C"
                    ]
                },
                "7": {
                    "types": [
                        "I"
                    ]
                },
                "5": {
                    "types": [
                        "I"
                    ]
                },
                "10": {
                    "types": [
                        "I"
                    ]
                },
                "2": {
                    "types": [
                        "I"
                    ]
                },
                "1": {
                    "types": [
                        "O"
                    ]
                },
                "9": {
                    "types": [
                        "C"
                    ]
                },
                "8": {
                    "types": [
                        "I"
                    ]
                },
                "3": {
                    "types": [
                        "I"
                    ]
                }
            }
        },
        "AG_4306": {
            "slots": {},
            "types": [
                "O",
                "I"
            ]
        },
        "P_4302_2": {
            "slots": {},
            "types": [
                "O",
                "I"
            ]
        },
        "AG_4307": {
            "slots": {},
            "types": [
                "O",
                "I"
            ]
        },
        "RIO1_Rack1": {
            "slots": {
                "1": {
                    "types": [
                        "C"
                    ]
                },
                "4": {
                    "types": [
                        "C"
                    ]
                },
                "2": {
                    "types": [
                        "O"
                    ]
                },
                "3": {
                    "types": [
                        "I"
                    ]
                }
            }
        },
        "AG_4302": {
            "slots": {},
            "types": [
                "I",
                "O"
            ]
        }
    }

**Comment by ottowayi** [comment]
`get_module_info` uses your current CIP route to the device, except it replaces the last hop with a `backplane/<slot>` segment.  The CIPDriver uses what you pass in literally, the LogixDriver will append a route to slot 0 if you just supply an IP address. The data you're showing from the LogixDriver comes from the tag database, it doesn't go out and search for it at all.  Can you share some details about how your devices are connected? If possible, a screenshot from RSLinx would be most helpful (my email is on my profile if you can't share it here)

**Comment by rpabech** [comment]
Here you go: 

![1](https://user-images.githubusercontent.com/29556434/195130575-466edace-82bb-483b-a137-aec96857970e.png)

![2](https://user-images.githubusercontent.com/29556434/195130592-5e0f11e2-a07a-4988-b8f2-ca58a8ce977c.png)

What I am looking for is how to collect the Slot info (IO Cards version, model, etc),

Thanks

Ricardo

**Reply by ottowayi** [reply]
Can you try opening a `CIPDriver` to `<New_INX_IO ip address>/bp/0` and then call `get_module_info` for each slot number?  That should work.  I apologize for the late response, I would normally test this out and show you the results, but I'm in the middle of a move at the moment and my devices are offline still.

**Comment by rpabech** [comment]
I will try that. But shouldn't this get_module_info work for the Local Modules on the Controller?

For the New_INX_IO those are a separate backplane, but I would expect the local one to work fine when opening a CIPDriver to the controller itself.

Thanks

**Reply by ottowayi** [reply]
yeah, it should be the same for both.  Is it when you try the rack with the controller that you get the same data for every slot?

**Comment by rpabech** [comment]
Yes but it does get the product code correctly. So itnseems that some data is being collected but not the rest. 

{'vendor': 'Rockwell Automation/Allen-Bradley', 'product_type': 'General Purpose Discrete I/O', 'product_code': 82, 'revision': {'major': 3, 'minor': 1}, 'status': b'a\x00', 'serial': '00000000', 'product_name': ''}
{'vendor': 'Rockwell Automation/Allen-Bradley', 'product_type': 'General Purpose Discrete I/O', 'product_code': 85, 'revision': {'major': 3, 'minor': 1}, 'status': b'a\x00', 'serial': '00000000', 'product_name': ''}


As you can see product name and revision and serial are not correct. Product code does change based on the IO model. 

Thanks

**Reply by ottowayi** [reply]
hmm that is weird, there are no _defaults_ for those values so that's what it's getting back from the device.  Can you use RSLinx and right-click, select `Device Properties`?  Both that dialog and this method are parsing the Identity object, so the data should match.  If it does not, then we know something with our request is wrong.

**Comment by rpabech** [comment]
Sure let me do that and get back to you!

Thanks
